tests/php/_inc/lib/markdown/test-class-wpcom-ghf-markdown-parser.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ * GitHub-Flavoured Markdown unit tests.
+ *
+ * @package Jetpack
+ */
+
+// Require the whole lib to process text.
+require_once JETPACK__PLUGIN_DIR . '_inc/lib/markdown/0-load.php';
+
+/**
+ * Class for testing the WPCom_GHF_Markdown_Parser class.
+ *
+ * @covers WPCom_GHF_Markdown_Parser
+ */
+class WP_Test_WPCom_GHF_Markdown_Parser extends WP_UnitTestCase {
+	/**
+	 * Test that links are preserved when going through the Markdown parser.
+	 *
+	 * @covers WPCom_GHF_Markdown_Parser->transform
+	 * @dataProvider get_text_urls
+	 *
+	 * @since 9.2.0
+	 *
+	 * @param string $text     The Markdown text we want to transform.
+	 * @param string $expected Expected HTML content.
+	 */
+	public function test_urls_preserve( $text, $expected ) {
+		/*
+		 * Text always ends with a newline.
+		 * Let's add it here (and not in the data provider)
+		 * to make things clearer there.
+		 */
+		$expected .= "\n";
+
+		$transformed_text = ( new WPCom_GHF_Markdown_Parser() )->transform( $text );
+		$this->assertEquals( $expected, $transformed_text );
+	}
+
+	/**
+	 * Get link examples to test how Markdown avoids transforming elements in links.
+	 *
+	 * @return array The test data.
+	 */
+	public function get_text_urls() {
+		return array(
+			'no_link_bold'          => array(
+				'Some **bold** text',
+				'Some <strong>bold</strong> text',
+			),
+			'link_bold'             => array(
+				'**[A bold link](https://jetpack.com/)**',
+				'<strong><a href="https://jetpack.com/">A bold link</a></strong>',
+			),
+			'link_undercore'        => array(
+				'[A link with underscore in URL](https://jetpack.com/_features_/)',
+				'<a href="https://jetpack.com/_features_/">A link with underscore in URL</a>',
+			),
+			'link_alone'            => array(
+				'https://jetpack.com/',
+				'https://jetpack.com/',
+			),
+			'link_underscore_alone' => array(
+				'https://jetpack.com/_features_/',
+				'https://jetpack.com/_features_/',
+			),
+			'ftp_link_underscore'   => array(
+				'ftp://_best_pack@jetpack.com:123',
+				'ftp://_best_pack@jetpack.com:123',
+			),
```

Comment:
Additional test cases to consider:
```
// This is current behavior before this patch.
'explicit URL' => array(
    'url: <https://jetpack.com/_features_/>',
    'url: <a href="https://jetpack.com/_features_/">https://jetpack.com/_features_/</a>',
),
// This too.
'URL in link text' => array(
    'url: [https://jetpack.com/](https://jetpack.com/#xyz)',
    'url: <a href="https://jetpack.com/#xyz">https://jetpack.com/</a>',
),

'bolded URL' => array(
    '**https://jetpack.com/_features_/**',
    '???',
),
'emphasized URL' => array(
    '_https://jetpack.com/_features_/_',
    '???',
),
'URL with odd but legal characters' => array(
    'https://example.com/_form?type=*&q=a+(b+or+c)&arr[]=1&arr[]=2&value=_foo_bar'
    '???',
),
```

Reply:
Ah, good catch. This complicates things quite a bit. I'm not quite sure how to catch all those with a single regex, to be honest. I'm not sure if this means we should reconsider the alternate approach, #17761, which only targets Twitter links. 

What do you think? Can you think of a way to address all those at once?

Reply:
It's a little hacky, but I think we can get most of the way there with something like
```
#(?<!<)(?:https?|ftp)://([^\s<>"\[\]()]+|\[(?1)*+\]|\((?1)*+\))+(?<![_*.?])#i
```
Explanation:
* The `(?<!<)` at the beginning avoids matching URLs immediately preceded by a `<`.
* The `(?<![_*.?!])` at the end avoids including any trailing `_`, `*`, `.`, `?`, or `!` in the URL.
* The `([^\s<>"\[\]()]+|\[(?1)*+\]|\((?1)*+\))+` is some PCRE magic that allows only balanced `[]` and `()` in the URL. So a URL like `https://example.com/foo(bar` would get cut off after "foo", while the whole of `https://example.com/foo(bar)` would be matched.
    * `[^\s<>"\[\]()]+` is the basic set of allowed characters.
    * `\[(?1)*+\]` allows an open-bracket, then any number of repetitions of anything capturing group 1 might match, then a close-bracket. The `*+` will hopefully avoid too much backtracking in case of an unmatched `[`.
    * `\((?1)*+\)` does the same for parens.

There are some edge cases, but hopefully they're rare enough that we can ignore them:
* `<http://example.com/_foo_` won't be protected.
* In something like `_emphasized text http://example.com/_foo_ more emphasized text_`, that trailing `_` won't be protected and it'll confuse Markdown into ending the emphasis early.
* Something like `http://example.com/(/_foo_` won't have the part after the unbalanced bracket/paren protected.

Reply:
I committed your suggestions in b6f529e6ccc90b70bfe9d5f194f01966d132d168